src/averageAges.js

```diff
@@ -55,10 +77,36 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const mothers = people.filter(mother => people.some(person =>
+    person.mother === mother.name));
+
+  let kids = people.filter(mummy => mummy.mother != null);
+
+  if (onlyWithSon === true) {
+    kids = people.filter(son => son.mother != null && son.sex === 'm');
+  }
+
+  let ageSum = 0;
+  let numberOfPairs = 0;
+
+  kids.forEach(kid => {
+    const mother = mothers.find(mummy => mummy.name === kid.mother);
+
+    if (mother) {
+      const ageDifference = kid.born - mother.born;
+
+      ageSum += ageDifference;
+      numberOfPairs++;
+    }
+  });
+
+  if (numberOfPairs > 0) {
+    const averageAgeDiff = ageSum / numberOfPairs;
+
+    return parseFloat(averageAgeDiff.toFixed(2));
+  } else {
+    return 0; // Zabezpieczenie przed dzieleniem przez zero
```

Comment:
The function name and its parameter name 'onlyWithSon' might be misleading. It suggests that you're filtering for sons only, but in the function, you're actually checking if children have mothers. Consider renaming to 'filterByParentalStatus' or something similar.